VT event lifecycle docs (#5425)
withastro/docs: src/content/docs/en/guides/view-transitions.mdx

Module scripts are only ever executed once because the browser keeps track of which modules are already loaded. For these scripts, you do not need to worry about re-execution.

## Lifecycle events

The `<ViewTransition />` router fires a number events on the `document` during navigation. These events provide hooks into the lifecycle of navigation, allowing to you do things like showing indicators that a new page is loading, override default behavior, and restore state as navigation is completing.

The navigation process involves a **preparation** phase, when new content is loaded; a **DOM swap** phase, where the old page's content is replaced by the new page's content; and a **completion** phase where the new page is loaded, scripts are executed and clean-up work is carried out. 

Astro's View Transition API lifecycle events in order are:
- [`astro:before-preparation`](#astrobefore-preparation)
- [`astro:after-preparation`](#astroafter-preparation)
- [`astro:before-swap`](#astrobefore-swap)
- [`astro:after-swap`](#astroafter-swap)
- [`astro:page-load`](#astropage-load)

> While some actions can be trigged during any event, some tasks can only be performed during a specific event for best results, such as displaying a loading spinner before preparation or overriding animation pairs before swapping content.

Reviewer: ```suggestion
:::tip
`before-` events allow you to influence and modify actions that are about to take place, and `after-` events are notifications that a phase is complete.
:::

While some actions can be triggered during any event, some tasks can only be performed during a specific event for best results, such as displaying a loading spinner before preparation or overriding animation pairs before swapping content.
```
Wondering what we think about something like this? I'll commit so we can see it on the page.

Author: Yes, I like this!

Module scripts are only ever executed once because the browser keeps track of which modules are already loaded. For these scripts, you do not need to worry about re-execution.

## Lifecycle events

> The `<ViewTransition />` router fires a number events on the `document` during navigation. These events provide hooks into the lifecycle of navigation, allowing you to do things like showing indicators that a new page is loading, override default behavior, and restore state as navigation is completing.

Reviewer: ```suggestion
The `<ViewTransition />` router fires a number events on the `document` during navigation. These events provide hooks into the lifecycle of navigation, allowing you to do things like show indicators that a new page is loading, override default behavior, and restore state as navigation is completing.
```
nit: consistent verb tense

    const originalLoader = ev.loader;
    ev.loader = async function() {
      const { startSpinner } = await import('./spinner.js');
      const stop = startSpinner();
      await originalLoader();
      stop();
    };
  });
</script>
```

### `astro:after-preparation`

<Since v="3.6.0" />

An event that fires at the end of the preparation phase, after the new page's content has been loaded and parsed into a document. This event occurs before the view transitions phase.

This example uses the `astro:before-preparation` event to start a loading indicator and the `astro:after-preparation` event to stop it:

```astro
<script is:inline>
  document.addEventListener('astro:before-preparation', () => {
    document.querySelector('#loading').classList.add('show');
  });
  document.addEventListener('astro:after-preparation', () => {
    document.querySelector('#loading').classList.remove('show');
  });
</script>
```
Compared to the spinner example above, this example looks simpler: If all of the listener's code can be executed synchronously, there is no need to hook into the loader's callback.  
### `astro:before-swap`

<Since v="3.6.0" />

> An event that fires befores the new document, populated during the preparation phase, replaces the current document. This event occurs inside of the view transition, where the user is still seeing a snapshot of the old page.

Reviewer: ```suggestion
An event that fires before the new document (which is populated during the preparation phase) replaces the current document. This event occurs inside of the view transition, where the user is still seeing a snapshot of the old page.
```
I feel like the parenthesis can help with the more complicated sentence structure here.

This event can be used to make changes before the swap occurs. The `newDocument` property on the event represents the incoming document. Here is an example of ensuring the browser's light or dark mode preference in `localStorage` is carried over to the new page:

```astro
<script is:inline>
  function setDarkMode(document) {
    let theme = localStorage.darkMode ? 'dark' : 'light';
    document.documentElement.dataset.theme = theme;
  }

  setDarkMode(document);

  document.addEventListener('astro:before-swap', ev => {
    // Pass the incoming document to set the theme on it
    setDarkMode(ev.newDocument);
  });
</script>
```

> 
> 
> The `astro:before-swap` event can also be used to change the *implementation* of the swap. The default swap implementation diffs head content, moves __persistent__ elements from the old document to the `newDocument`, and then replaces the entire `body` with the body of the new document.
> 
> At this point of the lifecycle, you could choose to define your own swap implementation, for example to diff the entire contents of the existing document (which some other routers do):

Reviewer: ```suggestion
The `astro:before-swap` event can also be used to change the *implementation* of the swap. The default swap implementation diffs head content, moves __persistent__ elements from the old document to the `newDocument`, and then replaces the entire `body` with the body of the new document.

At this point of the lifecycle, you could choose to define your own swap implementation, for example to diff the entire contents of the existing document (which some other routers do):
```
just removed some blank lines!

### `astro:before-preparation`

<Since v="3.6.0" />

An event that fires at the beginning of the preparation phase, after navigation has started (for example, after the user has clicked a link), but before content is loaded.

This event is useful for a few use-cases:

- To do something before loading has started, such as showing an a loading spinner.
- To alter loading, such as loading content you've defined in a template rather than from the external URL.
- To change the `direction` of the navigation (which is usually either `forward` or `backward`) in order to do custom animation.

Here's an example of using the `astro:before-preparation` event to load a spinner before the content is loaded.

```js
<script is:inline>
  document.addEventListener('astro:before-preparation', ev => {
    const originalLoader = ev.loader;
    ev.loader = async function() {
      const { startSpinner } = await import('./spinner.js');
      const stop = startSpinner();
      await originalLoader();
      stop();
    };
  });
</script>
```

> 

Reviewer 2: I like this example where the spinner is set and removed with the loader callback.    But if there is another wrapper around it that needs some time to do e.g. further prefetching based on the loaded data, then the spinner is gone too soon. It would be much better to remove the spinner in astro:after-preparation, because all loaders/prefetchers/... are through and the screenshot has not yet been taken.

Reviewer: Is this easy to show in a code sample? Do we just show **starting** the spinner here? I will ask you or Matthew to edit the sample to your liking!

Reviewer: @martrapp Given @matthewp 's new example below using both `before-` and `after-` preparation, should we remove this example altogether, and let the combined example below serve for both?

Module scripts are only ever executed once because the browser keeps track of which modules are already loaded. For these scripts, you do not need to worry about re-execution.

## Lifecycle events

The `<ViewTransition />` router fires a number events on the `document` during navigation. These events provide hooks into the lifecycle of navigation, allowing to you do things like showing indicators that a new page is loading, override default behavior, and restore state as navigation is completing.

> The navigation phase involves a **preparation** phase, when new content is loaded; a **DOM swap** phase, where the old page's content is replaced by the new page's content; and a **completion** phase where the new page is loaded, scripts are executed and clean-up work is carried out. 

Reviewer: ```suggestion
The navigation proccess involves a **preparation** phase, when new content is loaded; a **DOM swap** phase, where the old page's content is replaced by the new page's content; and a **completion** phase where the new page is loaded, scripts are executed and clean-up work is carried out. 
```

This event, when listened to on the **outgoing page**, is useful to pass along and restore any state on the DOM that needs to transfer over to the new page.

Previously this event was used to implement dark mode. You could listen to the event and then add the dark mode class name to, for example: `<html class="dark-mode">`. This is still a valid way to do it, but it might be better to do so in an earlier event, like `astro:after-preparation` which occurs before the swap and before the view transition occurs.

> As the `astro:after-swap` event occurs immediately before the view transition ends, it's a good time to stop any loading indicators that might be running.

Reviewer 2: "occurs immediately before the view transition ends" could be misleading: Surely it is near the end of the navigation, but "astro:after-swap" takes place immediately before the start of the animation (for the simulation immediately before the start of the "new" part). In any case, the loading indicator is already recorded on the screenshot and will not disappear so quickly, depending on the animation. A better place to remove it would be `astro:after-preparation` because the photo has not yet been taken :)

Reviewer: I'm hearing two different comments here:

1. "immediately before the view transition ends" isn't *exactly* right. What's a better, concise, way to describe it?
2. Spinner has come back to haunt us. This now is the only real example (that's not, it's still safe to do this) example for this section. Can you suggest (code sample even better, but just in words like above is fine) what a good thing to happen here is? So that this section can have a proper example of its own that makes sense and is a realistic use case? Note: it's also OK to say that "there's very little that might target this event specifically, but it does provide you..."

This event, when listened to on the **outgoing page**, is useful to pass along and restore any state on the DOM that needs to transfer over to the new page.

This is the latest point in the lifecyle where it is still safe to, for example, add a dark mode class name (`<html class="dark-mode">`), though you may wish to do so in an earlier event.

The "astro:after-swap" event occurs immediately after the browser history has been updated and the scroll position has been set. 

> 

Reviewer 2: ```suggestion
Therefore, it is save now to override the standard scroll restoration if desired. The following example resets the horizontal and vertical scroll position to the top left corner of the page. 
```

Reviewer 2: ```astro
document.addEventListener('astro:after-swap', 
  () => window.scrollTo({ left: 0, top: 0, behavior: 'instant' }))
```

Reviewer: Perfect, thanks, @martrapp  I'll add these in!

Just because GitHub is unintuitive and you might not realize you can add the code directly in your suggestion, here's my trick:

When you click to make a "suggestion", it will automatically format with three backticks. If you **increase the number of backticks to open and close your suggestions** (I add three, to make it six, out of habit), then you can have a regular three-backtick code sample directly in your suggestion! Like this!

![image](https://github.com/withastro/docs/assets/5098874/f0783b15-02ed-4109-80c8-e6aeade4786d)

This event, when listened to on the **outgoing page**, is useful to pass along and restore any state on the DOM that needs to transfer over to the new page.

This is the latest point in the lifecyle where it is still safe to, for example, add a dark mode class name (`<html class="dark-mode">`), though you may wish to do so in an earlier event.

The "astro:after-swap" event occurs immediately after the browser history has been updated and the scroll position has been set. 

> Therefore, it is now possible to override the default scroll restore for history navigation if desired. The following example resets the horizontal and vertical scroll position to the top left corner of the page for each navigation. 

Reviewer: ```suggestion
Therefore, one use of targeting this event is to override the default scroll restore for history navigation. The following example resets the horizontal and vertical scroll position to the top left corner of the page for each navigation. 
```

This event, when listened to on the **outgoing page**, is useful to pass along and restore any state on the DOM that needs to transfer over to the new page.

This is the latest point in the lifecyle where it is still safe to, for example, add a dark mode class name (`<html class="dark-mode">`), though you may wish to do so in an earlier event.

> As the `astro:after-swap` event occurs immediately before the view transition ends, it's a good time to stop any loading indicators that might be running.

Reviewer 2: Meaningful example pending
